metaworker/metaworker.go

```diff
@@ -17,50 +15,71 @@ import (
 	"github.com/moby/buildkit/exporter"
 	imageexporter "github.com/moby/buildkit/exporter/containerimage"
 	localexporter "github.com/moby/buildkit/exporter/local"
-	"github.com/moby/buildkit/frontend"
-	"github.com/moby/buildkit/frontend/dockerfile"
-	"github.com/moby/buildkit/frontend/gateway"
 	"github.com/moby/buildkit/session"
 	"github.com/moby/buildkit/snapshot/blobmapping"
 	"github.com/moby/buildkit/source"
 	"github.com/moby/buildkit/source/containerimage"
 	"github.com/moby/buildkit/source/git"
 	"github.com/moby/buildkit/source/local"
+	"github.com/moby/buildkit/worker"
+	digest "github.com/opencontainers/go-digest"
 )
 
-type pullDeps struct {
-	Snapshotter  ctdsnapshot.Snapshotter
-	ContentStore content.Store
-	Applier      diff.Differ
-	Differ       diff.Differ
-	Images       images.Store
+// MetaWorkerOpt is a constructor for MetaWorker
+type MetaWorkerOpt struct {
+	Name            string
+	MetadataStore   *metadata.Store
+	Worker          worker.Worker
+	BaseSnapshotter ctdsnapshot.Snapshotter
+	ContentStore    content.Store
+	Applier         diff.Differ
+	Differ          diff.Differ
+	ImageStore      images.Store
 }
 
-func defaultControllerOpts(root string, pd pullDeps) (*Opt, error) {
-	md, err := metadata.NewStore(filepath.Join(root, "metadata.db"))
-	if err != nil {
-		return nil, err
-	}
+// MetaWorker is a worker instance with dedicated snapshotter, cache, and so on.
+type MetaWorker struct {
+	MetaWorkerOpt
+	Snapshotter      ctdsnapshot.Snapshotter // blobmapping snapshotter
+	CacheManager     cache.Manager
+	SourceManager    *source.Manager
+	InstructionCache InstructionCache
+	Exporters        map[string]exporter.Exporter
+	SessionManager   *session.Manager
+	ImageSource      source.Source
+	CacheExporter    *cacheimport.CacheExporter
+	CacheImporter    *cacheimport.CacheImporter
+	// no frontend
+}
+
+type InstructionCache interface {
+	Probe(ctx context.Context, key digest.Digest) (bool, error)
+	Lookup(ctx context.Context, key digest.Digest, msg string) (interface{}, error) // TODO: regular ref
+	Set(key digest.Digest, ref interface{}) error
+	SetContentMapping(contentKey, key digest.Digest) error
+	GetContentMapping(dgst digest.Digest) ([]digest.Digest, error)
+}
 
-	snapshotter, err := blobmapping.NewSnapshotter(blobmapping.Opt{
-		Content:       pd.ContentStore,
-		Snapshotter:   pd.Snapshotter,
-		MetadataStore: md,
+func NewMetaWorker(opt MetaWorkerOpt) (*MetaWorker, error) {
```

Comment:
I'd call `metaWorker` `WorkerController` and add methods like `Add(name, worker),Get(constraint)` to it that the init would call and that could be later called when attaching remote workers.

Instead of `MetaWorkerOpt`, `runc.New(cliopts)` could return a worker interface that the `Add()` accepts. To keep this shared code there could be and util pkg with a function that sets up the default configuration by converting a struct.

Reply:
What name do you suggest for the type (`X`) that will be used for `WorkerController.Add(name string, worker X) error` and `WorkerController.Get(constraint Constraint) (X, error)`?

It will be a structure composed of a "bare" worker, a snapshotter, and some other stuffs. 
(as in the `MetaWorker` structure in the current WIP PR, but without sessionmgr and so on)

Reply:
Also, let me know whether it is the right approach to create `X` instances for each of snapshotters. (`containerd-overlay`, `containerd-btrfs`, `containerd-blahblahblah` instance rather than the single `containerd` instance)

I guess this model simplifies the design for local workers, but on second thought this might result in creating annoying number of remote worker instances.

Reply:
Is the correct way that that `X` is called a `Worker` (current /worker could become /executor) and has methods for implementing https://github.com/moby/buildkit/blob/3e4200c5fbdd7ebcac89ff1a136598de39ebe25d/solver/solver.go#L67-L76 for any `pb.Op` the solver/manager may provide and then internals like snapshotters are not exposed at all? If this is correct then maybe call it `OpWorker` first and then name switches in separate PR.

For the second question, my understanding is that containerd will have introspection API that can provide us these values. I think we should support all that containerd supports but not sure if we want to do this automatically/by default. Even if we show them all we should only enable one automatically as otherwise having more snapshotters will make builds slower(still have the same CPU but need to copy data between drivers). I think it would also be fine to just add the default one and if users want more they can enable them with cli flags.